## ตัวอย่าง Naive Bayes 

In [1]:
import pandas as pd
import numpy as np

In [2]:
# สร้าง dataframe ชื่อ data ว่างๆ ขึ้นมาก่อน
data = pd.read_csv('play_tennis.csv')
print(data)

     outlook temperature humidity   windy play
0      sunny         hot     high    weak   no
1      sunny         hot     high  strong   no
2   overcast         hot     high    weak  yes
3      rainy        mild     high    weak  yes
4      rainy        cool   normal    weak  yes
5      rainy        cool   normal  strong   no
6   overcast        cool   normal  strong  yes
7      sunny        mild     high    weak   no
8      sunny        cool   normal    weak  yes
9      rainy        mild   normal    weak  yes
10     sunny        mild   normal  strong  yes
11  overcast        mild     high  strong  yes
12  overcast         hot   normal    weak  yes
13     rainy        mild     high  strong   no


In [5]:
# สร้าง dataframe ชื่อ person ว่างๆ ขึ้นมาก่อน
new = pd.DataFrame()

# สร้างข้อมูลที่เป็น new data 
new['outlook'] = ['rain']
new['temperature'] = ['hot']
new['humidity'] = ['high']
new['windy'] = ['weak']

# View the data 
new


outlook temperature humidity windy
0    rain         hot     high  weak

In [6]:
# นับจำนวน yes
n_yes = len(data[data['play'] == 'yes'])  #len(object) เป็น function ใช้นับจำนวน object

# นับจำนวน no
n_no = len(data[data['play'] == 'no']) 

# นับจำนวนที่เล่นทั้งหมด
total_play = len(data['play'])

print(n_yes,n_no,total_play)

9 5 14


In [7]:
n_sunny_yes = len(data[(data['outlook'] == 'sunny') & (data['play'] == 'yes')])
n_rain_yes = len(data[(data['outlook'] == 'rainy') & (data['play'] == 'yes')])
n_overcast_yes = len(data[(data['outlook'] == 'overcast') & (data['play'] == 'yes')])

n_hot_yes = len(data[(data['temperature'] == 'hot') & (data['play'] == 'yes')])
n_mild_yes = len(data[(data['temperature'] == 'mild') & (data['play'] == 'yes')])
n_cool_yes = len(data[(data['temperature'] == 'cool') & (data['play'] == 'yes')])

n_high_yes = len(data[(data['humidity'] == 'high') & (data['play'] == 'yes')])
n_normal_yes = len(data[(data['humidity'] == 'normal') & (data['play'] == 'yes')])

n_weak_yes = len(data[(data['windy'] == 'weak') & (data['play'] == 'yes')])
n_strong_yes = len(data[(data['windy'] == 'strong') & (data['play'] == 'yes')])

In [8]:
n_sunny_no = len(data[(data['outlook'] == 'sunny') & (data['play'] == 'no')])
n_rain_no = len(data[(data['outlook'] == 'rainy') & (data['play'] == 'no')])
n_overcast_no = len(data[(data['outlook'] == 'overcast') & (data['play'] == 'no')])

n_hot_no= len(data[(data['temperature'] == 'hot') & (data['play'] == 'no')])
n_mild_no = len(data[(data['temperature'] == 'mild') & (data['play'] == 'no')])
n_cool_no = len(data[(data['temperature'] == 'cool') & (data['play'] == 'no')])

n_high_no = len(data[(data['humidity'] == 'high') & (data['play'] == 'no')])
n_normal_no = len(data[(data['humidity'] == 'normal') & (data['play'] == 'no')])

n_weak_no = len(data[(data['windy'] == 'weak') & (data['play'] == 'no')])
n_strong_no = len(data[(data['windy'] == 'strong') & (data['play'] == 'no')])


In [9]:
p_yes = n_yes/total_play 
p_no = n_no/total_play 

p_sunny_yes = n_sunny_yes/n_yes
p_rain_yes = n_rain_yes/n_yes
p_overcast_yes = n_overcast_yes/n_yes

p_hot_yes = n_hot_yes/n_yes
p_mild_yes = n_mild_yes/n_yes
p_cool_yes = n_cool_yes/n_yes

p_high_yes = n_high_yes/n_yes
p_normal_yes = n_normal_yes/n_yes

p_weak_yes = n_weak_yes/n_yes
p_strong_yes = n_strong_yes/n_yes

print(p_rain_yes)

0.3333333333333333


In [10]:
p_sunny_no = n_sunny_no/n_no
p_rain_no = n_rain_no/n_no
p_overcast_no = n_overcast_no/n_no

p_hot_no = n_hot_no/n_no
p_mild_no = n_mild_no/n_no
p_cool_no = n_cool_no/n_no

p_high_no = n_high_no/n_no
p_normal_no = n_normal_no/n_no

p_weak_no = n_weak_no/n_no
p_strong_no = n_strong_no/n_no

print(p_rain_no)

0.4


In [11]:
p_rain_hot_high_weak_yes = p_rain_yes * p_hot_yes * p_high_yes * p_weak_yes * p_yes

p_rain_hot_high_weak_no = p_rain_no * p_hot_no * p_high_no * p_weak_no * p_no

print(p_rain_yes,p_hot_yes,p_high_yes,p_weak_yes,p_yes)
print(p_rain_no,p_hot_no,p_high_no,p_weak_no,p_no)

print('p_rain_hot_high_weak_yes = ', p_rain_hot_high_weak_yes)
print('p_rain_hot_high_weak_no = ', p_rain_hot_high_weak_no)
print('p_rain_hot_high_weak_no = ',  "%.4f" %p_rain_hot_high_weak_no ) #คล้าย %.2f Format ใน printf() ภาษา C programming 
print('p_rain_hot_high_weak_no = ', round(p_rain_hot_high_weak_no,4))


0.3333333333333333 0.2222222222222222 0.3333333333333333 0.6666666666666666 0.6428571428571429
0.4 0.4 0.8 0.4 0.35714285714285715
p_rain_hot_high_weak_yes =  0.010582010582010581
p_rain_hot_high_weak_no =  0.01828571428571429
p_rain_hot_high_weak_no =  0.0183
p_rain_hot_high_weak_no =  0.0183


ค่า P(play = no|new) (=0.0183) มีค่ามากกว่า P(play = yes|new) (=0.0105) ดังนั้นโมเดลจึงทำนายว่าข้อมูลตัวอย่างนี้มีค่าคลาส play = no ไม่เล่น tennis